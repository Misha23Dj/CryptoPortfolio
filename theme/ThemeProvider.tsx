import React, { createContext, useContext, useMemo, useState } from "react";
import { useColorScheme } from "react-native";
import { dark, light, type Theme } from "./colors";

type ThemeMode = "light" | "dark" | "system";

const ThemeCtx = createContext<Theme>(light);
const ThemeModeContext = createContext<{
  mode: ThemeMode;
  setMode: (m: ThemeMode) => void;
}>({
  mode: "system",
  setMode: () => {},
});

export const useTheme = () => useContext(ThemeCtx);
export const useThemeMode = () => useContext(ThemeModeContext);

type Props = { children: React.ReactNode };

const ThemeProvider: React.FC<Props> = ({ children }) => {
  const [mode, setMode] = useState<ThemeMode>("system");
  const scheme = useColorScheme();

  const effective = mode === "system" ? scheme ?? "light" : mode;
  const theme = useMemo(
    () => (effective === "dark" ? dark : light),
    [effective]
  );

  return (
    <ThemeModeContext.Provider value={{ mode, setMode }}>
      <ThemeCtx.Provider value={theme}>{children}</ThemeCtx.Provider>
    </ThemeModeContext.Provider>
  );
};

export default ThemeProvider;
