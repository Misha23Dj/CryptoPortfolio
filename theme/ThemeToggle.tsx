import React from "react";
import { Pressable, StyleSheet, Text, View } from "react-native";
import { useTheme, useThemeMode } from "./ThemeProvider";

const MODES = ["light", "dark", "system"] as const;
type Mode = (typeof MODES)[number];

const ThemeToggle: React.FC = () => {
  const theme = useTheme();
  const { mode, setMode } = useThemeMode();

  return (
    <View
      style={[
        styles.wrap,
        { borderColor: theme.border, backgroundColor: theme.card },
      ]}
    >
      {MODES.map((m) => {
        const selected = mode === m;
        return (
          <Pressable
            key={m}
            onPress={() => setMode(m as Mode)}
            style={[
              styles.chip,
              {
                backgroundColor: selected ? theme.text : "transparent",
                borderColor: theme.border,
              },
            ]}
          >
            <Text
              style={{
                color: selected ? theme.bg : theme.text,
                fontWeight: "600",
              }}
            >
              {m[0].toUpperCase() + m.slice(1)}
            </Text>
          </Pressable>
        );
      })}
    </View>
  );
};

const styles = StyleSheet.create({
  wrap: {
    flexDirection: "row",
    gap: 8,
    padding: 6,
    borderWidth: StyleSheet.hairlineWidth,
    borderRadius: 12,
    alignSelf: "flex-end",
    marginRight: 16,
    marginBottom: 8,
  },
  chip: {
    paddingVertical: 6,
    paddingHorizontal: 10,
    borderWidth: StyleSheet.hairlineWidth,
    borderRadius: 10,
  },
});

export default ThemeToggle;
