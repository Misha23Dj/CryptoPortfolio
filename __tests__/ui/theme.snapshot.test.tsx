import ThemeProvider, { useTheme } from "@/theme/ThemeProvider";
import ThemeToggle from "@/theme/ThemeToggle";
import { fireEvent, render } from "@testing-library/react-native";
import React from "react";
import { Text, View } from "react-native";

const Demo = () => {
  const theme = useTheme();
  return (
    <View style={{ backgroundColor: theme.bg, padding: 8 }}>
      <Text style={{ color: theme.text }}>Demo</Text>
      <ThemeToggle />
    </View>
  );
};

it("switches theme and updates snapshot", () => {
  const { getByTestId, toJSON } = render(
    <ThemeProvider>
      <Demo />
    </ThemeProvider>
  );
  expect(toJSON()).toMatchSnapshot();
  fireEvent.press(getByTestId("theme-toggle"));
  expect(toJSON()).toMatchSnapshot();
});
