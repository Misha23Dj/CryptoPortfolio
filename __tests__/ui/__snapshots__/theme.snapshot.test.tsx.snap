// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`switches theme and updates snapshot 1`] = `
<View
  style={
    {
      "backgroundColor": "#ffffff",
      "padding": 8,
    }
  }
>
  <Text
    style={
      {
        "color": "#0f172a",
      }
    }
  >
    Demo
  </Text>
  <View
    accessibilityRole="button"
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    hitSlop={8}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      [
        {
          "alignItems": "center",
          "backgroundColor": "transparent",
          "borderRadius": 12,
          "justifyContent": "center",
        },
      ]
    }
    testID="theme-toggle"
  >
    <ViewManagerAdapter_SymbolModule
      animated={false}
      colors={[]}
      name="moon.fill"
      size={24}
      style={
        [
          {
            "height": 24,
            "width": 24,
          },
          {
            "height": 24,
            "width": 24,
          },
        ]
      }
      tint={4279179050}
      tintColor="#0f172a"
      type="monochrome"
    />
  </View>
</View>
`;

exports[`switches theme and updates snapshot 2`] = `
<View
  style={
    {
      "backgroundColor": "#0b1220",
      "padding": 8,
    }
  }
>
  <Text
    style={
      {
        "color": "#e5e7eb",
      }
    }
  >
    Demo
  </Text>
  <View
    accessibilityRole="button"
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    hitSlop={8}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      [
        {
          "alignItems": "center",
          "backgroundColor": "transparent",
          "borderRadius": 12,
          "justifyContent": "center",
        },
      ]
    }
    testID="theme-toggle"
  >
    <ViewManagerAdapter_SymbolModule
      animated={false}
      colors={[]}
      name="sun.max.fill"
      size={24}
      style={
        [
          {
            "height": 24,
            "width": 24,
          },
          {
            "height": 24,
            "width": 24,
          },
        ]
      }
      tint={4293257195}
      tintColor="#e5e7eb"
      type="monochrome"
    />
  </View>
</View>
`;
